docs/get_machine.1
.\"Made with Love
.TH get_machine 1 "JULY 2019" Linux "User Manuals"
.SH NAME

get_machine - identify running hardware architecture

.SH SYNOPSIS
.na

get_machine [ --exact --override OVERRIDE --OS ]

.SH DESCRIPTION

get_machine exists to make simple shell scripts that need
to know that hardware architecture it is assuming or what
host operating system is being used.
.br

At its core is figure out the general hardware architecture
and return it as quickly as possible. Although it is sometimes
useful to return something different; which is why --override
exits and thus what ever is supplied by with it is returned.
Scripts that wish to expose that can leverage the environment
variable GET_MACHINE_FLAGS to allow efficient overriding in
their scripts.
.br
If one wishes for something more exact than x86 or amd64, the
option --exact will return more specific values like i686-pae.
.br
Should one desire to know the host operating system: --OS
.br
A word of warning; --override always will take top precedence
and can return anything you desire.

.SH EXAMPLES

Typically, get_machine will be in scripts used in bootstrapping
.br
# get_machine
.br
# get_machine --override "I am the very model of a modern major general"
.br
# get_machine ${GET_MACHINE_FLAGS}
.br

.SH COMPATIBILITY

get_machine is compatible with all Turing complete machines;
even the ones that try to be Turing complete -1

.SH AUTHORS
Jeremiah Orians <Jeremiah@pdp10.guru>

.SH COPYRIGHT
Copyright 2016-2019 Jeremiah Orians <Jeremiah@pdp10.guru>
.br
License GPLv3+.

.SH "SEE ALSO"
M1(1), hex2(1), blood-elf(1), kaem(1), syscalls(2)


CHANGELOG.org
## Copyright (C) 2017 Jeremiah Orians
## This file is part of mescc-tools
##
## mescc-tools is free software: you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation, either version 3 of the License, or
## (at your option) any later version.
##
## mescc-tools is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with mescc-tools.  If not, see <http://www.gnu.org/licenses/>.

* Current
** Added
Added support for AMD64 dwarf footers in blood-elf via --64
Added hex0 for i386 in NASM, M1 and hex0
Added hex1 for i386 in NASM, M1, hex1 and hex0
Added first generation AARCH64 elf header
Added hex2 for i386 in NASM, M1, hex2 and hex1
Added M0 for i386 in NASM, M1 and hex2
Added catm for i386 in NASM, M1 and hex0
Added support for EOF in line comments in hex2 and M1; thanks to markjenkins
Added prototype M1 Manpage
Added prototype hex2 Manpage
Added prototype blood-elf Manpage
Added prototype kaem Manpage
Added prototype get_machine Manpage
Added cc_x86 for AMD64 in NASM
Added cc_x86 for x86 in NASM and M1

** Changed

** Fixed
Removed duplicate in kaem's help
Fixed regression in M1 in regards to knight null padding

** Removed

* 0.6 - 2019-04-14
** Added
Added template ELF headers for ARM
Added initial support for ARM
Added official hex0 seed for AMD64
Added official hex1 seed for AMD64
Added support for <Number null padding in M1 files
Added support for < alignment padding
Added support for ^!label, ^@label and ^%label alignment calculations
Added standard line error messages to hex2-linker
Added standard line error messages to M1-macro
Added --exact and --override to get_machine
Added GET_MACHINE_FLAGS variable to tests to allow overriding
Backported M2-Planet's support for Octal and binary
Added support for *BSDs to test the results of the build using sha256.sh (thanks to ng0)
Added official hex2 seed for AMD64
Added NASM versions of AMD64 bootstrap pieces for improve introspection of bootstrap
Added kaem.run script supporting intial phase of AMD64 bootstrap
Added catm C prototype to eliminate the dependency on cat and the shell supporting >
Added catm NASM prototype to simplify build
Added catm M1 prototype to reduce bootstrap dependency
Added catm hex0 prototype to eliminate bootstrap dependencies down to hex0
Added M0 NASM prototype to simplify build
Added M0 M1 prototype to reduce bootstrap dependency
Added M0 hex2 prototype to eliminate bootstrap dependencies down to hex2
Verified ARM port to support M2-Planet

** Changed
Updated build.sh and kaem.run to the current mescc-tools syntax
Reduced get_machine's build dependencies
Cleaned up x86 elf headers
Removed kaem's dependence on getopt
Replaced --Architecture with --architecture
changed get_machine's default output to filter machine names into known families
Reduced M1 null padding of strings to a single null for all architectures except Knight
Updated AMD64 bootstrap kaem.run to include steps from hex0 to M0

** Fixed
Fixed broken test9 thanks to janneke
Fixed wrong displacement calculations for ARM immediates
Fixed typo in license header
Fixed kaem.run to actually function and produce identical results
Fixed regression caused by linux 4.17
Removed false newline added in numerate_number for zero case
Fixed broken bootstrap script

** Removed
Removed final dependency on getopt
Removed need to know architecture numbers as that was a bad idea

* 0.5 - 2018-06-15
** Added
Added INSTALL notes
Added HACKING notes
Added examples of minimal Hex1, Hex2 and M1-macro programs that may need to be
written to bootstrap a particular architecture.
Added useful functions to reduce bootstrap dependencies
Added support for binary output in M1-macro

** Changed
Changed Knight architecture offset calculation to match new standard
Updated test3 lisp.s to include more functionality
Updated test3 definitions file to reflect changes in Knight instruction encoding
enhanced README to be more useful
Pulled numerate_string functionality out of hex2 and M1 into a shared library
Eliminated getopt from M1-Macro, hex2-linker and blood-elf; use --Architecture 1 instead of --Architecture=1

** Fixed
Corrected M1-macro incorrectly expressing negative numbers
Updated test3 checksum to reflect new version of lisp.s
fixed check.sh to actually perform all checks.
Fixed build.sh to function in a self-hosting fashion

** Removed
Removed blood-elf's dependency on getopt
Removed C preprocessor macro from blood-elf needed for mescc support
Removed hex2's dependency on getopt
Removed C preprocessor macro from hex2 needed for mescc support
Removed need for octal support in the building of hex2
Removed M1's dependency on getopt
Removed C preprocessor macro from M1 needed for mescc support
Removed need for sprintf from M1

* 0.4 - 2018-02-24
** Added
Added file checks to reduce the number of error messageless faults
Added a current generation M1.M1 file as a test for mescc-tools
Added prototype kaem build script
M1-macro now catches undefined macros to allow easier troubleshooting
Added kaem build tool
Added ability to track build progress in kaem
Added support for line escapes in kaem
Added support for --strict in kaem to halt in the event of errors
Added selectable script file support in kaem
Added support for PATH search to kaem with fallbacks in the event of NULL environments

** Changed
flipped blood-elf from ignoring :: to :_
converted test8 into a full test
Added bash style line comments to kaem
Added support for raw strings to kaem
Stopped showing comment lines in kaem --verbose
Removed dependence on getenv to have more control over environmental lookup

** Fixed

get_machine.c
/* -*- c-file-style: "linux";indent-tabs-mode:t -*- */
/* Copyright (C) 2017 Jeremiah Orians
 * This file is part of mescc-tools.
 *
 * mescc-tools is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mescc-tools is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mescc-tools.  If not, see <http://www.gnu.org/licenses/>.
 */


#include <stdio.h>
#include <stdlib.h>
#include <sys/utsname.h>
void file_print(char* s, FILE* f);
int match(char* a, char* b);

#define TRUE 1
//CONSTANT TRUE 1
#define FALSE 0
//CONSTANT FALSE 0

/* Standard C main program */
int main(int argc, char **argv)
{
	int exact = FALSE;
	int override = FALSE;
	char* override_string;
	int option_index = 1;

	struct utsname* unameData = calloc(1, sizeof(struct utsname));
	uname(unameData);

	while(option_index <= argc)
	{
		if(NULL == argv[option_index])
		{
			option_index = option_index + 1;
		}
		else if(match(argv[option_index], "--exact"))
		{
			exact = TRUE;
			option_index = option_index + 1;
		}
		else if(match(argv[option_index], "--override"))
		{
			override = TRUE;
			override_string = argv[option_index + 1];
			option_index = option_index + 2;
		}
		else if(match(argv[option_index], "--OS"))
		{
			file_print(unameData->sysname, stdout);
			fputc('\n', stdout);
			exit(EXIT_SUCCESS);
		}
		else if(match(argv[option_index], "-V") || match(argv[option_index], "--version"))
		{
			file_print("get_machine 0.6.0\n", stdout);
			exit(EXIT_SUCCESS);
		}
		else if(match(argv[option_index], "-h") || match(argv[option_index], "--help"))
		{
			file_print("If you want exact architecture use --exact\n", stderr);
			file_print("If you want to know the Operating system use --OS\n", stderr);
			file_print("If you wish to override the output to anything you want use --override\n", stderr);
			exit(EXIT_SUCCESS);
		}
		else
		{
			file_print("Unknown option\n", stderr);
			exit(EXIT_FAILURE);
		}
	}

	if(override) file_print(override_string, stdout);
	else if(!exact)
	{
		if(match("i386", unameData->machine) ||
		   match("i486", unameData->machine) ||
		   match("i586", unameData->machine) ||
		   match("i686", unameData->machine) ||
		   match("i686-pae", unameData->machine)) file_print("x86", stdout);
		else if(match("x86_64", unameData->machine)) file_print("amd64", stdout);
		else file_print(unameData->machine, stdout);
	}
	else file_print(unameData->machine, stdout);
	file_print("\n", stdout);
	return EXIT_SUCCESS;
}
